### PR TITLE
feat(write): tab-aware writes, insert command, multi-tab safety (0.7.1)

### PR DESCRIPTION
**Pull request by LucaDeLeo**
## Summary

Adds the missing CLI surface for populating tabs in multi-tab docs and stops `write` from silently collapsing them. Closes the biggest footgun in the `pull`/`write` asymmetry.

### New commands

- **`gdoc insert DOC --tab NAME FILE`** — populate a specific tab from a local markdown file. Works on empty tabs (previously impossible: `add-tab` + `edit --tab` couldn't find an anchor in an empty body). Strips YAML frontmatter. `--position start|end` controls insertion point.
- **`gdoc write --tab NAME`** — scoped write replacing one tab's body via the Docs API; siblings untouched.

### Behavior changes

- **`gdoc write`** refuses to collapse multi-tab docs into one. Exits 3 with a message pointing at `--tab NAME`, `gdoc insert`, or the new `--force-collapse-tabs` opt-in. The legacy collapsing behavior is still available, just no longer silent.
- **`gdoc write`** strips YAML frontmatter from the input before upload. Pull→edit→write round-trips no longer dump visible YAML into the doc body.
- **`gdoc edit --old-file FILE`** is now usable alone → deletes the matched range. `--new-file` alone still errors and points at `gdoc insert`.
- **`gdoc add-tab`** prints a clickable `?tab=...` URL alongside the bare tabId.

### Fixes

- `replace_formatted` skips zero-width `deleteContentRange` requests (Docs API rejects empty ranges with `"The range should not be empty"`).
- `__version__` resynced with `pyproject.toml` (drift to 0.6.3 → 0.7.1).

### New helpers in `gdoc.api.docs`

- `insert_markdown_into_tab(doc_id, tab_name, markdown, position, replace)` — encapsulates the tab-resolve + parse_markdown + batchUpdate dance that users previously had to write by hand against the Python API.
- `count_document_tabs(doc_id)` — lightweight tab count using a `fields="tabs(tabProperties/tabId,childTabs)"` mask so the per-write safety check doesn't download tab body content.

### Release mechanics

- CHANGELOG.md added, linked from README.
- `gdoc update` notice and post-update message both link to the changelog URL.
- Version bumped to 0.7.1.

## Test plan

- [x] `uv run pytest tests/ -v` — 832 passing (21 new tests: insert handler, tab-aware write, multi-tab safety, zero-width replace, count_document_tabs, frontmatter strip, add-tab URL).
- [x] `uv run ruff check` — no new issues introduced.
- [x] `bash scripts/check-no-stubs.sh` — clean.
- [x] Smoke test end-to-end against a real Google Doc: `gdoc new` → `add-tab` (URL printed) → `insert --tab` on the empty tab → `cat --tab` verified bullets/bold/code render → `write` on multi-tab doc refused with correct exit 3 → `write --tab` scoped replacement → sibling tab untouched → frontmatter round-trip stripped → `--force-collapse-tabs` opt-in collapses as expected.

See [CHANGELOG.md](./CHANGELOG.md) for the full release notes.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * New gdoc insert command for inserting markdown into specific tabs; per-tab write support; CLI now shows clickable tab URLs.
  * CLI now points to changelog when reporting or completing updates.

* **Bug Fixes**
  * Safer find-and-replace behavior to avoid empty-range errors.
  * Version number synchronized across packaging.

* **Documentation**
  * Added comprehensive CHANGELOG and README changelog link.

* **Tests**
  * Expanded test coverage for tab operations, inserts, writes, and edge cases.

### CODE REVIEW

File: gdoc/api/docs.py [hunk 1]
```diff
@@ -124,6 +124,35 @@ def get_document_tabs(doc_id: str) -> list[dict]:
...
+    the call is cheap enough to run as a per-write safety check.
+    """
+    try:
+        service = get_docs_service()
+        doc = (
+            service.documents()
+            .get(
+                documentId=doc_id,
+                fields="tabs(tabProperties/tabId,childTabs)",
+            )
```
[the remark this answers is not in the document]
Comment (LucaDeLeo) [reply gdoc/api/docs.py:143]: Fixed in 206a4da. You were right — `includeTabsContent=True` is required for `Document.tabs` to be populated. The fields mask alone wasn't enough (empirical evidence to the contrary in my earlier smoke test was misleading, probably from a stale build). Added the flag while keeping the minimal fields mask, and the unit test now asserts both.

File: gdoc/cli.py [hunk 2]
```diff
@@ -694,27 +796,56 @@ def cmd_write(args) -> int:
     except OSError as e:
         raise GdocError(f"cannot read file: {e}", exit_code=3)
 
+    # Strip frontmatter — pull prepends it, and leaving it in the upload
+    # dumps visible YAML into the doc body.
+    from gdoc.frontmatter import parse_frontmatter
+    _, content = parse_frontmatter(content)
```
[the remark this answers is not in the document]
Comment (LucaDeLeo) [reply gdoc/cli.py:801]: Fixed in 206a4da. Tightened `parse_frontmatter` to only strip the leading `---...---` block when at least one `key: value` line parses out of it — innocent markdown with a thematic break is now round-tripped unchanged. Other callers (`cmd_push`, `cmd_ci`, `cmd_pull_hook`) already gate on `"gdoc" in metadata` and fall through either way, so the change is behavior-preserving for them. Added a regression test for the thematic-break-plus-`---`-footer case.

File: gdoc/cli.py [hunk 3]
```diff
@@ -333,6 +338,95 @@ def cmd_add_tab(args) -> int:
...
+    if not os.path.isfile(file_path):
+        raise GdocError(f"file not found: {file_path}", exit_code=3)
+    try:
+        with open(file_path) as f:
+            content = f.read()
+    except OSError as e:
+        raise GdocError(f"cannot read file: {e}", exit_code=3)
+
+    from gdoc.frontmatter import parse_frontmatter
+    _, content = parse_frontmatter(content)
```
[the remark this answers is not in the document]
Comment (LucaDeLeo) [reply gdoc/cli.py:399]: Already fixed at the parser level in 206a4da — `parse_frontmatter` now requires at least one `key: value` line to treat a leading `---...---` block as metadata, so `cmd_insert` (which uses the same parser) also preserves thematic-break content. Added a regression test (`test_thematic_break_with_prose_not_stripped`) that covers both call sites.

File: gdoc/cli.py [hunk 4]
```diff
@@ -333,6 +339,95 @@ def cmd_add_tab(args) -> int:
...
+    position = getattr(args, "position", "start")
+    file_path = args.file
+
+    if not os.path.isfile(file_path):
+        raise GdocError(f"file not found: {file_path}", exit_code=3)
+    try:
+        with open(file_path) as f:
+            content = f.read()
+    except OSError as e:
+        raise GdocError(f"cannot read file: {e}", exit_code=3)
```
[the remark this answers is not in the document]
Comment (LucaDeLeo) [reply gdoc/cli.py]: Fixed in 7560a15 — added `from e` to the OSError re-raise in both `cmd_insert` (line 396) and `cmd_write` (line 796). Left the four pre-existing occurrences at lines 600/931/1133/1602 alone since they're untouched by this PR; happy to clean those up in a follow-up if you'd like.

File: gdoc/cli.py [hunk 5]
```diff
@@ -1931,6 +2081,34 @@ def build_parser() -> GdocArgumentParser:
...
+        "--tab", required=True,
+        help="Target tab by title or ID",
+    )
+    insert_p.add_argument(
+        "--position", choices=["start", "end"], default="start",
+        help="Insert at the start (default) or end of the tab body",
+    )
+    insert_p.add_argument(
+        "--force", action="store_true", help="Skip conflict detection"
+    )
```
[the remark this answers is not in the document]
Comment (LucaDeLeo) [reply gdoc/cli.py:2106]: Fixed in 7560a15 — help text is now "Proceed even if the doc changed since the last read", matching your suggested phrasing.

File: tests/test_write.py [hunk 6]
```diff
@@ -14,7 +14,12 @@
 
 
 def _make_args(**overrides):
-    """Build a SimpleNamespace mimicking parsed write args."""
+    """Build a SimpleNamespace mimicking parsed write args.
+
+    Defaults `force_collapse_tabs=True` so the safety check doesn't
+    require patching `get_document_tabs` in every legacy test. Tests
+    that exercise the multi-tab safety path should override it.
+    """
```
[the remark this answers is not in the document]
Comment (LucaDeLeo) [reply tests/test_write.py:24]: Fixed in 7560a15. Good catch — `_make_args()` now defaults `force_collapse_tabs=False` to match `build_parser()`. Rather than adding `@patch` boilerplate to every success-path test, I added a module-level autouse fixture that stubs `count_document_tabs` → `1`. `TestWriteCollapseSafety`'s own `@patch` decorators stack via LIFO so the inner patches win when the collapse test needs 2 tabs. 39/39 write tests pass.
